tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/SparseTensor.java

```diff
@@ -0,0 +1,163 @@
+ * @param <T> type of data stored in the tensor
+ */
+public interface SparseTensor<T extends TType> extends Tensor {
+
+  /**
+   * Creates a sparse tensor from {@code indices}, {@code values} and {@code denseShape} dense
+   * tensors.
+   *
+   * @param indices A 2-D tensor of shape {@code [N, ndims]}, that specifies the indices of the
+   *     elements in the sparse tensor that contain non-default values (elements are zero-indexed).
+   *     For example, {@code indices=[[1,3,1], [2,4,0]]} specifies that the elements with indexes of
+   *     {@code [1,3,1]} and {@code [2,4,0]} have non-default values.
+   * @param values A 1-D tensor of shape {@code [N]}, which supplies the values for each element in
+   *     indices. For example, given {@code indices=[[1,3,1], [2,4,0]]}, the parameter {@code
+   *     values=[18, 3.8]} specifies that element {@code [1,3,1]} of the sparse tensor has a value
+   *     of {@code 18}, and element {@code [2,4,0]} of the tensor has a value of {@code 3.8}.
+   * @param denseShape A 1-D tensor of shape {@code [ndims]} where each the value at index {@code i}
+   *     represents the size of dimension {@code i} in a dense version of that tensor.
+   * @return the new sparse tensor
+   * @throws IllegalArgumentException if shapes of the dense tensors are not compatible
+   */
+  static <T extends TType> SparseTensor<T> of(TInt64 indices, T values, TInt64 denseShape) {
+    if (indices.rank() != 2) {
+      throw new IllegalArgumentException("Sparse indices must be a rank-2 tensor");
+    }
+    if (values.rank() != 1) {
+      throw new IllegalArgumentException("Sparse values must be a rank-1 tensor");
+    }
+    if (denseShape.rank() != 1) {
+      throw new IllegalArgumentException("Sparse shape must be a rank-1 tensor");
+    }
+    if (indices.shape().get(0) != values.shape().get(0)) {
+      throw new IllegalArgumentException(
+          "Number of indices must be equal to the number of values ["
+              + indices.shape().get(0)
+              + " != "
+              + values.shape().get(0)
+              + "]");
+    }
+    if (indices.shape().get(1) != denseShape.shape().get(0)) {
+      throw new IllegalArgumentException(
+          "Indices must have a coordinate for each dimensions of the tensor ["
+              + indices.shape().get(1)
+              + " != "
+              + denseShape.shape().get(0)
+              + "]");
+    }
+    // Use mapper of the values tensor as this is the one giving the type of the sparse tensor as
+    // well
+    TensorMapper<T> mapper = (TensorMapper<T>) values.asRawTensor().typeInfo().mapper();
+
+    // Keep a strong reference to all sub-tensors of the sparse tensor
+    try (PointerScope scope = new PointerScope()) {
+      scope.attach(indices.asRawTensor().nativeHandle());
+      scope.attach(values.asRawTensor().nativeHandle());
+      scope.attach(denseShape.asRawTensor().nativeHandle());
+      return mapper.mapSparse(indices, values, denseShape, scope);
```

Comment:
This only works because `mapSparse` calls extend on the scope right? In which case given this code is internal to our library, is it better to not use the scope in a try with resources and then remove the extend?

The fact that the scope looks like it's closed when you first read this method is quite confusing.

Reply:
I understand your concerns but this technique is used all over the place in our codebase (ex. [here](https://github.com/tensorflow/java/blob/1f691925187017346b19ec944a6ca6ebd83d2451/tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/RawTensor.java#L149)). This is to prevent memory leaks in case something goes wrong before we can store a strong reference to the scope in an autocloseable object.

For example, if you attach the `indices` to the scope but fail to attach the `values` (let say that `values` has been closed before somehow), this will throw an exception but will leave the `indices` attached to the "phantom" scope that has not been closed properly and leave its reference count inaccurate. In such case, closing the `indices` later won't release it until that phantom scope is garbage collected (I think, not sure about this part).

In other words, this is extra safety we use everywhere in case something unexpected happens.

Reply:
done